clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,190 @@
+use std::ops::ControlFlow;
+
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{path_to_local, path_to_local_id};
+use rustc_ast::{LitKind, Mutability};
+use rustc_errors::Applicability;
+use rustc_hir::{Block, Expr, ExprKind, HirId, LetStmt, Node, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::adjustment::{Adjust, AutoBorrow, AutoBorrowMutability};
+use rustc_session::declare_lint_pass;
+use rustc_span::{Span, sym};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the use of `seq.is_empty()` in an if-conditional where `seq` is a slice, array, or Vec,
+    /// and in which the first element of the sequence is indexed.
+    ///
+    /// ### Why is this bad?
+    /// This code is unnecessarily complicated and can instead be simplified to the use of an
+    /// if..let expression which accesses the first element of the sequence.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if !a.is_empty() {
+    ///    let b = a[0];
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a: &[i32] = &[1];
+    /// if let Some(b) = a.first() {
```

Comment:
Note that you can generate even simplier expressions:

- If `a` is a slice ref, use: `if let [b, ..] = a`
- If `a` is an array, use: `if let [b, ..] = &a`
- If `a` derefs to a slice, use: `if let [b, ..] = &a[..]`

Reply:
I can add it as an alternate suggestion, which one is more preferable though?

Reply:
They're in order of preference. `a` is the simplest, `&a` is needed if `a` is `[T]` or `[T; _]`, and `&a[..]` is needed for everything else.